src/mpd.c

```diff
@@ -3193,6 +3212,291 @@ mpd_command_update(struct evbuffer *evbuf, int argc, char **argv, char **errmsg,
+
+      /* |:todo:| real sticker implementation */
+      *errmsg = safe_asprintf("no such sticker");
+      ret = ACK_ERROR_NO_EXIST;
+      goto exit;
+
+    case MPD_STICKER_DELETE:
+      if (name_is_rating)
+	{
+	  if (ctx->rating_readonly)
+	    {
+	      *errmsg = safe_asprintf("%s is read-only", ctx->rating_name);
+	      ret = ACK_ERROR_PERMISSION;
+	      goto exit;
+	    }
+	  rating = 0;
+	  set_rating = 1;
+	  break;
+	}
+
+      /* |:todo:| real sticker implementation */
+      *errmsg = safe_asprintf("no such sticker");
+      ret = ACK_ERROR_NO_EXIST;
+      goto exit;
+
+    case MPD_STICKER_LIST:
+      if (mfi && mfi->rating)
+	{
+	  rating = mfi->rating / ctx->rating_factor;
+	  evbuffer_add_printf(evbuf, "sticker: %s=%d\n", ctx->rating_name, rating);
+	}
+
+      /* |:todo:| real sticker implementation */
+      break;
+
+      /* |:todo:| MPD_STICKER_FIND */
+    case MPD_STICKER_FIND:
+      if (name_is_rating && argc == 5)
+	{
+	  struct query_params qp;
+	  struct db_media_file_info dbmfi;
+	  char *c1;
+
+	  memset(&qp, 0, sizeof(struct query_params));
+
+	  qp.type = Q_ITEMS;
+	  qp.sort = S_VPATH;
+	  qp.idx_type = I_NONE;
+
+	  c1 = db_mprintf("(f.path LIKE '%s%%' AND f.rating != 0)", path);
+	  if (!c1)
+	    {
+	      *errmsg = safe_asprintf("Out of memory");
+	      ret = ACK_ERROR_UNKNOWN;
+	      goto exit;
+	    }
+	  qp.filter = c1;
+
+	  ret = db_query_start(&qp);
+	  if (ret < 0)
+	    {
+	      db_query_end(&qp);
+	      free(qp.filter);
+
+	      *errmsg = safe_asprintf("Could not start query");
+	      ret = ACK_ERROR_UNKNOWN;
+	      goto exit;
+	    }
+
+	  while (((ret = db_query_fetch_file(&qp, &dbmfi)) == 0) && (dbmfi.id))
+	    {
+	      ret = safe_atou32(dbmfi.rating, (unsigned int *) &rating);
+	      if (ret < 0)
+		{
+		  DPRINTF(E_LOG, L_MPD, "Error %s=%s doesn't convert to integer, song id: %s\n",
+		    ctx->rating_name, dbmfi.rating, dbmfi.id);
+		  continue;
+		}
+	      rating /= ctx->rating_factor;
+	      if (!rating)
+		/* inconsistent database */
+		continue;
+
+	      ret = evbuffer_add_printf(evbuf,
+		"file: file:%s\n"
+		"sticker: %s=%d\n",
+		   dbmfi.path,
+		   ctx->rating_name,
+		   rating);
+	      if (ret < 0)
+		DPRINTF(E_LOG, L_MPD, "Error adding song to the evbuffer, song id: %s\n", dbmfi.id);
+	    }
+	  ret = 0;
+
+	  db_query_end(&qp);
+	  free(qp.filter);
+	  break;
+	}
+
+      *errmsg = safe_asprintf("bad request");
+      ret = ACK_ERROR_ARG;
+      goto exit;
+
+    default:
+      *errmsg = safe_asprintf("bad request");
+      ret = ACK_ERROR_ARG;
+      goto exit;
+    }
+
+  if (mfi && set_rating && mfi->rating != rating)
+    {
+      DPRINTF(E_DBG, L_MPD, "STICKER notification for changed rating: %d -> %d\n", mfi->rating, rating);
+      mfi->rating = rating;
+      /* Note, that a DATABASE event is triggered, but the
+       * client actually expects a sticker event, so add it here. */
+      mpd_add_idle_events(LISTENER_STICKER);
+      db_file_update(mfi);
```

Comment:
A new function in db.c would be better (e. g. `db_file_set_rating`). This function should then trigger the STICKER event after updating the db.

Reply:
@chme This is more of a hint and definitely needs more thought.
For MPD, if only the rating changes, there should not be a `changed: database` event, which triggers reloading of the entire database, only a `changed: sticker` event, which (e.g., in cantata) only triggers reloading of the rating stickers for the current playist.
However, I am not sure what the implications of suppressing the LISTENER_DATABASE event in other subsystems would be.
Intuitively, I would think that LISTENER_DATABASE should be split into LISTENER_DATABASE and LISTENER_STICKER. Clients should then (de-)multiplex as necessary.